File: components/micro-gateway-core/src/main/java/org/wso2/micro/gateway/core/throttle/ThrottleCounter.java
/*
 *  Copyright (c) 2020, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

package org.wso2.micro.gateway.core.throttle;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.TimeUnit;

/**
 * This class is responsible for maintaining the throttle counters for various throttle policies.
 */
public class ThrottleCounter {
    private static final Logger log = LoggerFactory.getLogger(ThrottleCounter.class);

    private static final Map<String, ThrottleData> resourceLevelCounter = new ConcurrentHashMap<>();
    private static final Map<String, ThrottleData> applicationLevelCounter = new ConcurrentHashMap<>();
    private static final Map<String, ThrottleData> subscriptionLevelCounter = new ConcurrentHashMap<>();

    public void updateCounters(String apiKey, String appKey, boolean stopOnQuota, String subscriptionKey,
            long appTierCount, long appTierUnitTime, String appTierTimeUnit, long apiTierCount, long apiTierUnitTime,
            long subscriptionTierCount, long subscriptionTierUnitTime, String subcriptionTierTimeUnit,
            String resourceKey, long resourceTierCount, long resourceTierUnitTime, String resourceTierTimeUnit,
            long timestamp) {
        updateMapCounters(resourceLevelCounter, resourceKey, stopOnQuota, resourceTierCount, resourceTierUnitTime,
                resourceTierTimeUnit, timestamp, ThrottleData.ThrottleType.RESOURCE);
        updateMapCounters(applicationLevelCounter, appKey, stopOnQuota, appTierCount, appTierUnitTime, appTierTimeUnit,
                timestamp, ThrottleData.ThrottleType.APP);
        updateMapCounters(subscriptionLevelCounter, subscriptionKey, stopOnQuota, subscriptionTierCount,
                subscriptionTierUnitTime, subcriptionTierTimeUnit, timestamp, ThrottleData.ThrottleType.SUBSCRIPTION);
    }

    private void updateMapCounters(Map<String, ThrottleData> counterMap, String throttleKey, boolean stopOnQuota,
            long limit, long unitTime, String timeUnit, long timestamp, ThrottleData.ThrottleType throttleType) {
        ThrottleData existingThrottleData = counterMap.computeIfPresent(throttleKey, (key, throttleData) -> {
            if (limit > 0 && throttleData.getCount().incrementAndGet() >= limit) {
                throttleData.setThrottled(true);
            } else {
                throttleData.setThrottled(false);
            }
            if (timestamp > throttleData.getWindowStartTime() + throttleData.getUnitTime()) {
                throttleData.getCount().set(1);
                long startTime = timestamp - (timestamp % getTimeInMilliSeconds(1, timeUnit));
                throttleData.setWindowStartTime(startTime);
                throttleData.setThrottled(false);
            }
            if (log.isDebugEnabled()) {
                log.debug("Throttle count for the key '" + throttleKey + "' is " + throttleData.getCount());
            }
            return throttleData;
        });
        if (existingThrottleData == null) {
            counterMap.computeIfAbsent(throttleKey, key -> {
                ThrottleData throttleData = new ThrottleData();
                long startTime = timestamp - (timestamp % getTimeInMilliSeconds(1, timeUnit));
                throttleData.setWindowStartTime(startTime);
                throttleData.setStopOnQuota(stopOnQuota);
                throttleData.setUnitTime(getTimeInMilliSeconds(unitTime, timeUnit));
                throttleData.setThrottleType(throttleType);
                throttleData.getCount().set(0);
                throttleData.setThrottleKey(key);
                ThrottleDataReceiver.getThrottleDataCleanUpTask().addThrottleData(throttleData);
                if (log.isDebugEnabled()) {
                    log.debug("Throttle key inserted " + throttleKey);
                }
                return throttleData;
            });
            //There can be scenarios where the two threads stops at computeIfAbsent and one thread adds it to the map
            //and the second thread will go without incrementing the count. This additional computation is done to avoid
            // that scenario
            counterMap.computeIfPresent(throttleKey, (key, throttleData) -> {
                throttleData.getCount().incrementAndGet();
                return throttleData;
            });
        }
    }

    static boolean isResourceThrottled(String resourceKey) {
        return isRequestThrottled(resourceLevelCounter, resourceKey);
    }

    static boolean isAppLevelThrottled(String appKey) {
        return isRequestThrottled(applicationLevelCounter, appKey);
    }

    static boolean isSubscriptionLevelThrottled(String subscriptionKey) {
        return isRequestThrottled(subscriptionLevelCounter, subscriptionKey);
    }

    static void removeFromResourceCounterMap(String key) {
        resourceLevelCounter.remove(key);
    }

    static void removeFromApplicationCounterMap(String key) {
        applicationLevelCounter.remove(key);
    }

    static void removeFromSubscriptionCounterMap(String key) {
        subscriptionLevelCounter.remove(key);
    }

    private static boolean isRequestThrottled(Map<String, ThrottleData> counterMap, String throttleKey) {
        if (counterMap.containsKey(throttleKey)) {
            long currentTime = System.currentTimeMillis();
            ThrottleData throttleData = counterMap.get(throttleKey);
            if (currentTime > throttleData.getWindowStartTime() + throttleData.getUnitTime()) {
                counterMap.computeIfPresent(throttleKey, (key, throttleData1) -> {
                    throttleData1.setThrottled(false);
                    return throttleData1;
                });
                if (log.isDebugEnabled()) {
                    log.debug("Throttle window has expired. CurrentTime : " + currentTime + "\n Window start time : "
                            + throttleData.getWindowStartTime() + "\n Unit time : " + throttleData.getUnitTime());
                }
                return false;
            }
            return throttleData.isThrottled();
        }
        return false;
    }

    private long getTimeInMilliSeconds(long unitTime, String timeUnit) {
        long milliSeconds;
        if ("min".equalsIgnoreCase(timeUnit)) {
            milliSeconds = TimeUnit.MINUTES.toMillis(unitTime);
        } else if ("hour".equalsIgnoreCase(timeUnit)) {
            milliSeconds = TimeUnit.HOURS.toMillis(unitTime);
        } else if ("day".equalsIgnoreCase(timeUnit)) {
            milliSeconds = TimeUnit.DAYS.toMillis(unitTime);
        } else if ("week".equalsIgnoreCase(timeUnit)) {
            milliSeconds = 7 * TimeUnit.DAYS.toMillis(unitTime);
        } else if ("month".equalsIgnoreCase(timeUnit)) {
            milliSeconds = 30 * TimeUnit.DAYS.toMillis(unitTime);
        } else if ("year".equalsIgnoreCase(timeUnit)) {
            milliSeconds = 365 * TimeUnit.DAYS.toMillis(unitTime);
        } else {
            throw new RuntimeException("Unsupported time unit provided");
        }
        return milliSeconds;
    }
}

File: filter-chain/filters/throttleCounter.go
package filters

import (
	_ "github.com/golang/protobuf/ptypes/timestamp"
	log "github.com/sirupsen/logrus"
	"strconv"
	"strings"
	"sync"
	"time"
)

var apiLevelCounter map[string]ThrottleData
var resourceLevelCounter map[string]ThrottleData
var applicationLevelCounter map[string]ThrottleData
var subscriptionLevelCounter map[string]ThrottleData

func initiateCounters() {
	apiLevelCounter = make(map[string]ThrottleData)
	resourceLevelCounter = make(map[string]ThrottleData)
	subscriptionLevelCounter = make(map[string]ThrottleData)
	applicationLevelCounter = make(map[string]ThrottleData)
}

type ThrottleEvent struct {
	key          string
	stopOnQuota  bool
	TierCount    int64
	TierUnitTime int64
	TierTimeUnit string
	timestamp    int64
	throttleType ThrottleType
}

func (dataReferenceObject DataReference) updateCounters(mtx *sync.Mutex) {
	apiData := ThrottleEvent{
		key:          dataReferenceObject.apiKey,
		stopOnQuota:  dataReferenceObject.stopOnQuota,
		TierCount:    dataReferenceObject.apiTierCount,
		TierUnitTime: dataReferenceObject.apiTierUnitTime,
		TierTimeUnit: dataReferenceObject.apiTierTimeUnit,
		timestamp:    dataReferenceObject.timestamp,
		throttleType: ThrottleType(3),
	}
	resourceData := ThrottleEvent{
		key:          dataReferenceObject.resourceKey,
		stopOnQuota:  dataReferenceObject.stopOnQuota,
		TierCount:    dataReferenceObject.resourceTierCount,
		TierUnitTime: dataReferenceObject.resourceTierUnitTime,
		TierTimeUnit: dataReferenceObject.resourceTierTimeUnit,
		timestamp:    dataReferenceObject.timestamp,
		throttleType: ThrottleType(2),
	}
	appData := ThrottleEvent{
		key:          dataReferenceObject.appKey,
		stopOnQuota:  dataReferenceObject.stopOnQuota,
		TierCount:    dataReferenceObject.appTierCount,
		TierUnitTime: dataReferenceObject.appTierUnitTime,
		TierTimeUnit: dataReferenceObject.resourceTierTimeUnit,
		timestamp:    dataReferenceObject.timestamp,
		throttleType: ThrottleType(0),
	}
	subData := ThrottleEvent{
		key:          dataReferenceObject.subscriptionKey,
		stopOnQuota:  dataReferenceObject.stopOnQuota,
		TierCount:    dataReferenceObject.subscriptionTierCount,
		TierUnitTime: dataReferenceObject.subscriptionTierUnitTime,
		TierTimeUnit: dataReferenceObject.subscriptionTierTimeUnit,
		timestamp:    dataReferenceObject.timestamp,
		throttleType: ThrottleType(1),
	}

	mtx.Lock()
	apiLevelCounter = updateMapCounters(apiLevelCounter, apiData)
	mtx.Unlock()
	mtx.Lock()
	resourceLevelCounter = updateMapCounters(resourceLevelCounter, resourceData)
	mtx.Unlock()
	mtx.Lock()
	applicationLevelCounter = updateMapCounters(applicationLevelCounter, appData)
	mtx.Unlock()
	mtx.Lock()
	subscriptionLevelCounter = updateMapCounters(subscriptionLevelCounter, subData)
	mtx.Unlock()
}

func updateMapCounters(counterMap map[string]ThrottleData, event ThrottleEvent) map[string]ThrottleData {
	throttleKey := event.key
	stopOnQuota := event.stopOnQuota
	limit := event.TierCount
	unitTime := event.TierUnitTime
	timeUnit := event.TierTimeUnit
	timestamp := event.timestamp
	throttleType := event.throttleType

	throttleData, found := counterMap[throttleKey]
	if found {
		count := throttleData.getCount() + 1
		throttleData.setCount(count)
		if limit > 0 && count >= limit {
			throttleData.setThrottled(true)
		} else {
			throttleData.setThrottled(false)
		}
		if timestamp > (throttleData.getWindowStartTime() + throttleData.getUnitTime()) {
			throttleData.setCount(1)
			var startTime int64 = timestamp - (timestamp % getTimeInMilliSeconds(1, timeUnit))
			throttleData.setWindowStartTime(startTime)
			throttleData.setThrottled(false)
		}
		log.Infof("Throttle count for the key" + throttleKey + "is " +
			strconv.FormatInt(int64(throttleData.getCount()), 10))
		counterMap[throttleKey] = throttleData
		return counterMap
	} else {
		var throttleData ThrottleData = ThrottleData{}
		var startTime int64 = timestamp - (timestamp % getTimeInMilliSeconds(1, timeUnit))
		throttleData.setWindowStartTime(startTime)
		throttleData.setStopOnQuota(stopOnQuota)
		throttleData.setUnitTime(getTimeInMilliSeconds(unitTime, timeUnit))
		throttleData.setThrottleType(throttleType)
		throttleData.setCount(0)
		throttleData.setThrottleKey(throttleKey)
		addThrottleData(throttleData)
		counterMap[throttleKey] = throttleData
		return counterMap
	}
}

func isApiLevelThrottled(apiKey string) bool {
	return isRequestThrottled(apiLevelCounter, apiKey)
}

func isResourceThrottled(resourceKey string) bool {
	return isRequestThrottled(resourceLevelCounter, resourceKey)
}

func isAppLevelThrottled(appKey string) bool {
	return isRequestThrottled(applicationLevelCounter, appKey)
}

func isSubsLevelThrottled(subscriptionKey string) bool {
	return isRequestThrottled(subscriptionLevelCounter, subscriptionKey)
}

func removeFromResourceCounterMap(key string) {
	delete(resourceLevelCounter, key)
}

func removeFromApplicationCounterMap(key string) {
	delete(applicationLevelCounter, key)
}

func removeFromApiCounterMap(key string) {
	delete(apiLevelCounter, key)
}

func removeFromSubscriptionCounterMap(key string) {
	delete(subscriptionLevelCounter, key)
}

func isRequestThrottled(counterMap map[string]ThrottleData, throttleKey string) bool {
	if _, found := counterMap[throttleKey]; found {
		var currentTime int64 = getCurrentTimeMillis()
		var throttleData ThrottleData = counterMap[throttleKey]
		if currentTime > throttleData.getWindowStartTime()+throttleData.getUnitTime() {
			throttleData.setThrottled(false)
			counterMap[throttleKey] = throttleData
			log.Println("Throttle window has expired.")
			return false
		}
		return throttleData.isThrottled()
	}
	return false
}

func getCurrentTimeMillis() int64 {
	now := time.Now()
	unixNano := now.UnixNano()
	umillisec := unixNano / 1000000
	return umillisec
}

func getTimeInMilliSeconds(unitTime int64, timeUnit string) int64 {
	var milliSeconds int64
	if strings.EqualFold("min", timeUnit) {
		milliSeconds = time.Minute.Milliseconds() * unitTime
	} else if strings.EqualFold("hour", timeUnit) {
		milliSeconds = time.Hour.Milliseconds() * unitTime
	} else if strings.EqualFold("day", timeUnit) {
		milliSeconds = 24 * time.Hour.Milliseconds() * unitTime
	} else if strings.EqualFold("week", timeUnit) {
		milliSeconds = 7 * 24 * time.Hour.Milliseconds() * unitTime
	} else if strings.EqualFold("month", timeUnit) {
		milliSeconds = 30 * 24 * time.Hour.Milliseconds() * unitTime
	} else if strings.EqualFold("year", timeUnit) {
		milliSeconds = 365 * 24 * time.Hour.Milliseconds() * unitTime
	} else {
		log.Warnf("Unsupported time unit provided")
	}
	return milliSeconds
}
